gh-gl-sync/SpackCIBridge.py

```diff
@@ -163,6 +164,112 @@ def build_local_branches(self, open_prs):
 
+    def make_status_for_pipeline(self, pipeline):
+        """Generate POST data to create a GitHub status from a GitLab pipeline
+           API response
+        """
+        post_data = {}
+        if "status" not in pipeline:
+            return post_data
+
+        if pipeline["status"] == "created":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline has been created"
+
+        elif pipeline["status"] == "waiting_for_resource":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline is waiting for resources"
+
+        elif pipeline["status"] == "preparing":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline is preparing"
+
+        elif pipeline["status"] == "pending":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline is pending"
+
+        elif pipeline["status"] == "running":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline is running"
+
+        elif pipeline["status"] == "manual":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline is running manually"
+
+        elif pipeline["status"] == "scheduled":
+            post_data["state"] = "pending"
+            post_data["description"] = "Pipeline is scheduled"
+
+        elif pipeline["status"] == "failed":
+            post_data["state"] = "error"
+            post_data["description"] = "Pipeline failed"
+
+        elif pipeline["status"] == "canceled":
+            post_data["state"] = "failure"
+            post_data["description"] = "Pipeline was canceled"
+
+        elif pipeline["status"] == "skipped":
+            post_data["state"] = "failure"
+            post_data["description"] = "Pipeline was skipped"
+
+        elif pipeline["status"] == "success":
+            post_data["state"] = "success"
+            post_data["description"] = "Pipeline succeeded"
+
+        post_data["target_url"] = pipeline["web_url"]
+        post_data["context"] = "ci/gitlab-ci"
+
+        post_data = json.dumps(post_data).encode('utf-8')
+        return post_data
+
+    def dedupe_pipelines(self, api_response):
+        """Prune pipelines API response to only include the most recent result for each SHA"""
+        pipelines = {}
+        for response in api_response:
+            sha = response['sha']
+            if sha not in pipelines:
+                pipelines[sha] = response
+            else:
+                existing_datetime = dateutil.parser.parse(pipelines[sha]['updated_at'])
+                current_datetime = dateutil.parser.parse(response['updated_at'])
+                if current_datetime > existing_datetime:
+                    pipelines[sha] = response
+        return pipelines
+
+    def get_pipeline_api_template(self, gitlab_host, gitlab_repo):
+        template = gitlab_host.replace("ssh.", "https://")
```

Comment:
I'm not crazy about how specific this line is to our setup, and I think it follows from how the input arguments are defined.

After taking a closer look, I suggest the following:

 - "Repos" should be full URLs pointing to a repo that you would hand off to `git`, as-is, for cloning. (e.g.: `ssh://git@ssh.gitlab.spack.io/...`)

 - "Hosts" should be full `http(s)://` URLs pointing to the web host that is managing the corresponding "Repo". (e.g.: `https://gitlab.spack.io`)

   - We can continue to assume `https://github.com` for Github.

 - "Projects" should be the portion of a "Repo" that identifies the project managed by the given "Host". (e.g.: `scottwittenburg/gitlab-ci-tests` for the Github "Project" and `scott/gitlab-ci-tests` for the Gitlab "Project").

 - Given the above, change the CLI arguments to `<github_project> <gitlab_repo> <gitlab_host> <gitlab_project>`.

   - So, instead of `scottwittenburg/gitlab-ci-tests ssh.gitlab.spack.io scott/gitlab-ci-tests` (where you currently have to assume that the Gitlab host is given by replacing the leading `ssh.` in the `gitlab_repo` with `https://`), the script is called with `scottwittenburg/gitlab-ci-tests ssh://git@ssh.gitlab.spack.io/scott/gitlab-ci-tests https://gitlab.spack.io scott/gitlab-ci-tests` and there are no more ambiguities.